src/Compilers/CSharp/Portable/Symbols/Metadata/PE/TupleTypeDecoder.cs

```diff
@@ -0,0 +1,313 @@
+    ///     [TupleElementNamesAttribute(new[] { "e3", "e4", "e1", "e2" });
+    ///     class C : BaseType&lt;System.ValueTuple&lt;
+    ///         System.ValueTuple&lt;int,int&gt;, int&gt;
+    ///     { ... }
+    /// </code>
+    /// </example>
+    /// </summary>
+    internal struct TupleTypeDecoder
+    {
+        private readonly AssemblySymbol _containingAssembly;
+        private readonly ImmutableArray<string> _elementNames;
+        // Keep track of how many names we've "used" during decoding. Starts at
+        // the back of the array and moves forward.
+        private int _namesIndex;
+
+        private TupleTypeDecoder(ImmutableArray<string> elementNames, AssemblySymbol containingAssembly)
+        {
+            _containingAssembly = containingAssembly;
+            _elementNames = elementNames;
+            _namesIndex = elementNames.IsDefault ? 0 : elementNames.Length;
+        }
+
+        public static TypeSymbol DecodeTupleTypesIfApplicable(
+            TypeSymbol metadataType,
+            EntityHandle targetSymbolToken,
+            PEModuleSymbol containingModule)
+        {
+            Debug.Assert((object)metadataType != null);
+            Debug.Assert((object)containingModule != null);
+
+            ImmutableArray<string> elementNames;
+            // First check if the type is an interface. Interface implementations
+            // currently don't record tuple names in an attribute
+            // TODO(https://github.com/dotnet/roslyn/issues/12347)
+            if (metadataType.IsInterfaceType())
+            {
+                elementNames = default(ImmutableArray<string>);
+                var decoder = new TupleTypeDecoder(elementNames, containingModule.ContainingAssembly);
+                return decoder.DecodeNamedType((NamedTypeSymbol)metadataType);
+            }
+
+            if (containingModule.Module.HasTupleElementNamesAttribute(targetSymbolToken, out elementNames))
+            {
+                var decoder = new TupleTypeDecoder(elementNames, containingModule.ContainingAssembly);
+                try
+                {
+                    var decoded = decoder.DecodeType(metadataType);
+                    if (decoder._namesIndex == 0)
+                    {
+                        return decoded;
+                    }
+                    // If not all of the names have been used, the metadata is bad
+                }
+                catch (InvalidOperationException)
+                {
+                    // Indicates that the tuple info in the attribute didn't match
+                    // the type. Bad metadata.
+                }
+                // Bad metadata
+                return new UnsupportedMetadataTypeSymbol();
+            }
+
+            // No TupleElementNamesAttribute, check if it's a tuple nonetheless
+            if (metadataType.IsTupleCompatible())
+            {
+                return TupleTypeSymbol.Create((NamedTypeSymbol)metadataType);
+            }
```

Comment:
I did add a few tests for this in CIL (since that can't be properly defined in source right now), but I'll round it out to make sure this case is fully tested.